File: smart_travel_BE/src/main/java/com/example/smart_travel_BE/entity/Review.java
package com.example.smart_travel_BE.entity;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.NotFound;
import org.hibernate.annotations.NotFoundAction;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Data
@Entity
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Table(name = "reviews",
        indexes = {
                @Index(name = "idx_user_id", columnList = "user_id"),
                @Index(name = "idx_destination_id", columnList = "destination_id"),
                @Index(name = "idx_created_at", columnList = "created_at DESC")
        })
public class Review {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    @NotFound(action = NotFoundAction.IGNORE)
    private User user;  // Assuming User entity exists

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "destination_id")
    @JsonIgnore
    private Destination destination;

    @Column(name = "hotel_id")
    private Long hotelId;

    @Column(name = "tour_id")
    private Long tourId;

    @Column(name = "rating", nullable = false)
    private Integer rating;

    @Column(name = "comment", columnDefinition = "TEXT")
    private String comment;

    @Column(name = "likes_count", nullable = false)
    private Integer likesCount = 0;

    @Column(name = "is_approved", nullable = false)
    private Boolean isApproved = false;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;

    @OneToMany(mappedBy = "review", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<ReviewImage> images = new ArrayList<>();
}

File: smart_travel_FE/lib/domain/entities/review.dart
class Review{
  final int id; // BẮT BUỘC, không null
  final int? userId; // Chỉ lưu ID thay vì User object
  final String? userFullName; // Tên người dùng (lấy từ BE nếu có)
  final int? destinationId;
  final int? hotelId;
  final int? tourId;
  final int? rating;
  final String? comment;
  final int? likesCount;
  final bool? isApproved;
  final DateTime? createdAt;
  final DateTime? updatedAt;
  final List<String>? imageUrls; // Danh sách URL ảnh

  Review({
    required this.id,
    this.userId,
    this.userFullName,
    this.destinationId,
    this.hotelId,
    this.tourId,
    this.rating,
    this.comment,
    this.likesCount,
    this.isApproved,
    this.createdAt,
    this.updatedAt,
    this.imageUrls,
  });
}